Bond/react: molmap option

jrgissing opened the pull request:
**Summary**

Add molmap option to 'reset_mol_ids' keyword to provide more precise control over how to update molecule IDs, when not resetting them globally

In short, if the molecule IDs are defined consistently in the pre- and post-reaction templates (in the Molecules sections), the molmap option will ensure molecule IDs are updated consistently in the simulation. 

Details: The 'molmap' option uses the 'Molecules' section of both the pre- and post-reaction templates. Consider Atom 1 in the post-reaction template. If Atom 3 in the pre-reaction template is the first pre-reaction atom that has the same molecule ID as Atom 1, then the molecule ID of the simulation atom corresponding to Atom 1 is updated to the simulation molecule ID corresponding to Atom 3.

**Related Issue(s)**

none

**Author(s)**

Jake

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

<!--Please state whether any changes in the pull request will break backward compatibility for inputs, and - if yes - explain what has been changed and why-->

**Implementation Notes**

<!--Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected-->

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

jrgissing committed:
add more control to mol ID updates

jrgissing committed:
Merge branch 'lammps:develop' into bond/react-molmap_option

jrgissing committed:
copy/paste error in bond/react example

jrgissing committed:
Merge branch 'bond/react-molmap_option' of https://github.com/jrgissing/lammps into bond/react-molmap_option

jrgissing committed:
consistent 'nofix' options for write_data

'fix bond/react' writes some extra sections that parsers don't always know how to handle

jrgissing committed:
another reaction map validity check

jrgissing committed:
Merge branch 'develop' into bond/react-molmap_option

akohlmey commented:
@jrgissing this PR has been sitting idle for some time. Can you give us a status update?
If it won't be ready for a while, would you mind closing the request and resubmitting it when it is closer to being complete?

jrgissing commented:
I got back to this last week. Just working out a corner case, should be ready in a week or so

jrgissing committed:
better treatment for 'new mol IDs'

for molecule IDs in post-reaction template, but not in pre-reaction template, adds new molecule ID onto previous max  molecule ID

jrgissing committed:
initialize 'created atoms' mol ID to zero

can control using molmap option

jrgissing committed:
clarify how molecule IDs treated for

jrgissing committed:
molmap docs attempt

not a simple feature to describe in words

jrgissing committed:
initialization bug

akohlmey committed:
Merge branch 'develop' into bond/react-molmap_option

akohlmey committed:
better align with LAMMPS' programming conventions

akohlmey committed:
documentation fixes

akohlmey committed:
cosmetic, more LAMMPS style

akohlmey committed:
modernize some error messages
